miden-tx/src/testing/mock_chain.rs

```diff
@@ -0,0 +1,637 @@
+            nullifier_root,
+            note_root,
+            tx_hash,
+            proof_hash,
+            timestamp,
+        );
+
+        let block = Block::new(
+            header,
+            self.pending_objects.updated_accounts.clone(),
+            self.pending_objects.output_note_batches.clone(),
+            self.pending_objects.created_nullifiers.clone(),
+        )
+        .unwrap();
+
+        for (batch_index, note_batch) in self.pending_objects.output_note_batches.iter().enumerate()
+        {
+            for (note_index, note) in note_batch.iter().enumerate() {
+                // All note details should be OutputNote::Full at this point
+                match note {
+                    OutputNote::Full(note) => {
+                        let block_note_index = BlockNoteIndex::new(batch_index, note_index);
+                        let note_path = notes_tree.get_note_path(block_note_index).unwrap();
+                        let note_inclusion_proof = NoteInclusionProof::new(
+                            block.header().block_num(),
+                            block.header().sub_hash(),
+                            block.header().note_root(),
+                            block_note_index.to_absolute_index(),
+                            note_path,
+                        )
+                        .unwrap();
+
+                        self.available_notes.insert(
+                            note.id(),
+                            InputNote::authenticated(note.clone(), note_inclusion_proof),
+                        );
+                    },
+                    _ => continue,
+                }
+            }
+        }
+
+        for removed_note in self.removed_notes.iter() {
+            self.available_notes.remove(removed_note);
+        }
+
+        self.blocks.push(block.clone());
+        self.chain.add(header.hash());
+        self.reset_pending();
+
+        block
+    }
+
+    fn reset_pending(&mut self) {
+        self.pending_objects = PendingObjects::new();
+        self.removed_notes = vec![];
+    }
+
+    // ACCESSORS
+    // ----------------------------------------------------------------------------------------
+
+    /// Get the latest [ChainMmr].
+    pub fn chain(&self) -> ChainMmr {
+        let block_headers: Vec<BlockHeader> = self.blocks.iter().map(|b| b.header()).collect();
+        mmr_to_chain_mmr(&self.chain, &block_headers).unwrap()
+    }
+
+    /// Get a reference to [BlockHeader] with `block_number`.
+    pub fn block_header(&self, block_number: usize) -> BlockHeader {
+        self.blocks[block_number].header()
+    }
+
+    /// Get a reference to the nullifier tree.
+    pub fn nullifiers(&self) -> &Smt {
+        &self.nullifiers
+    }
+
+    pub fn available_notes(&self) -> Vec<InputNote> {
+        self.available_notes.values().cloned().collect()
+    }
+}
+
+pub struct MockChainBuilder {
+    accounts: Vec<Account>,
+    notes: Vec<Note>,
+    starting_block_num: u32,
+}
```

Comment:
The relation between `MockChain` and `MockChain` builder is a bit unclear to me. It probably makes sense to write some documentation about how these supposed to be used, but also, if we have a `MockChainBuilder`:

- Shouldn't all mutators on the `MockChain` be private to this module?
- Does `MockChain` need `pending_objects` field?

Basically, if we have a builder we usually add things to the builder and then build an object that doesn't need to be mutated further - but here it seems like the pattern is a bit different?

In general, I like the idea of having a separate builder - but I think if we go this way, we should move things like `add_nullifier()`, `add_new_wallet()`, `seal_block()` etc. to be on the builder rather than on the `MockChain`.